crates/subspace-farmer/src/plot.rs

```diff
@@ -363,9 +367,10 @@ impl IndexHashToOffsetDB {
             })
     }
 
-    fn put(&self, index: PieceIndex, offset: PieceOffset) -> io::Result<()> {
+    fn put(&self, index_hash: impl Into<PieceIndexHash>, offset: PieceOffset) -> io::Result<()> {
+        let distance = index_hash.into().distance(self.address);
```

Comment:
Should I just make a distance function to be a bitxor between piece index hash and address, so that it would be unique?

I don't like idea to store first and last piece cause that would make lookup algorithm more difficult at the moment it piece range shrinks. We would need to update range boundaries and do a search, while in current case we just pick the biggest key.

Reply:
> I don't like idea to store first and last piece cause that would make lookup algorithm more difficult at the moment it piece range shrinks. We would need to update range boundaries and do a search, while in current case we just pick the biggest key.

You don't need to store it, you just read the first value in a database and then the last one (right now you only read the last one). You can cache it in memory though to avoid disk lookups if plot doesn't change, but that is about it.

Also I believe your current distance doesn't account for overflow over the edge, in a sense that the things that overflow will be considered far away. Effectively you need to treat everything as 256-bit big-endian number (in binary) and use logic like here: https://github.com/subspace/subspace/blob/f59154ae1cd7d2df9b233c3b77f108f1d82bc3d1/crates/sp-consensus-subspace/src/digests.rs#L42-L45

In order to make the logic I described work, you need to offset keys such that in `0..2^256-1` domain `node_id` is at `2^255-1` or exactly in the middle (all keys should be offset like this), then you don't need to worry about overflow and can just use first/lask key in RocksDB.

Reply:
Changed to the xor metric instead. It should be fine here, as every piece index should yield different distance (it is bijection). There won't be any overflow also, and we can preserve only the notion of distance with max piece.